pallets/funding/src/runtime_api.rs

```diff
@@ -169,6 +173,36 @@ impl<T: Config> Pallet<T> {
 		ct_amount
 	}
 
+	pub fn get_next_vesting_schedule_merge_candidates(
+		account_id: AccountIdOf<T>,
+		hold_reason: <T as Config>::RuntimeHoldReason,
+		end_max_delta: Balance,
+	) -> Option<(u32, u32)> {
+		let schedules = pallet_linear_release::Vesting::<T>::get(account_id, hold_reason)
+			.expect("No vesting schedules found")
+			.to_vec();
+
+		let mut inspected_schedules: Vec<(usize, &pallet_linear_release::VestingInfoOf<T>)> = Vec::new();
+		let now = <frame_system::Pallet<T>>::block_number();
+		for (i, schedule) in schedules.iter().enumerate() {
+			if schedule.starting_block > now {
+				continue;
+			}
+
+			let schedule_end = schedule.ending_block_as_balance::<BlockNumberToBalanceOf<T>>();
+
+			for (j, other_schedule) in inspected_schedules.iter() {
+				let other_end = other_schedule.ending_block_as_balance::<BlockNumberToBalanceOf<T>>();
+				let end_delta = schedule_end.abs_diff(other_end);
+				if end_delta <= end_max_delta {
+					return Some((*j as u32, i as u32));
+				}
+			}
+			inspected_schedules.push((i, schedule));
+		}
+		None
+	}
+
```

Comment:
I focused mainly on this for now. I tried to perform few optimization, like removing the outer `for`, using a `BTreeMap` instead of a `Vec` and perform and early filtering using the `filter_map` directly on the schedules coming from the `Vesting` storage item in the Linear Release pallet. We have an hard upper limit of 100 `VestingInfo` per account, so it shouldn't change too much, and it's also executed in a RuntimeAPI, so it's not part of the consensus, but anyway, let me know what you think about these changes.

```rust
pub fn get_next_vesting_schedule_merge_candidates(
	account_id: AccountIdOf<T>,
	hold_reason: <T as Config>::RuntimeHoldReason,
	end_max_delta: Balance,
) -> Option<(u32, u32)> {
	let schedules = pallet_linear_release::Vesting::<T>::get(account_id, hold_reason)
		// TODO: Check if it is safe to use the default coming from `unwrap_or_default`.
		.unwrap_or_default()
		.into_iter()
		.enumerate()
		// Filter out schedules with future starting blocks before collecting them into a vector.
		.filter_map(|(i, schedule)| {
			if schedule.starting_block > <frame_system::Pallet<T>>::block_number() {
				None
			} else {
				Some((i, schedule.ending_block_as_balance::<BlockNumberToBalanceOf<T>>()))
			}
		})
		.collect::<Vec<_>>();

	let mut inspected_schedules = BTreeMap::new();

	for (i, schedule_end) in schedules {
		let range_start = schedule_end.saturating_sub(end_max_delta);
		let range_end = schedule_end.saturating_add(end_max_delta);

		//  All entries where the ending_block is between range_start and range_end.
		if let Some((_, &j)) = inspected_schedules.range(range_start..=range_end).next() {
			return Some((j as u32, i as u32));
		}

		inspected_schedules.insert(schedule_end, i);
	}

	None
}
```

Reply:
Looks good!
I just replaced the `unwrap_or_default()` with `?` for an early `None` return